controllers/secret_updater.go

```diff
@@ -7,96 +7,213 @@ import (
 	meowsv1alpha1 "github.com/cybozu-go/meows/api/v1alpha1"
 	"github.com/cybozu-go/meows/github"
+	"github.com/cybozu-go/meows/metrics"
+	"github.com/cybozu-go/well"
+	"github.com/go-logr/logr"
+	"github.com/prometheus/client_golang/prometheus"
 	corev1 "k8s.io/api/core/v1"
 	"k8s.io/apimachinery/pkg/types"
 	"sigs.k8s.io/controller-runtime/pkg/client"
-	"sigs.k8s.io/controller-runtime/pkg/log"
 )
 
-func NewSecretUpdater(client client.Client, interval time.Duration, githubClient github.Client) SecretUpdater {
-	return SecretUpdater{
+type metricsSet struct {
+	retryCount prometheus.Gauge
+}
+
+func newSecretUpdater(log logr.Logger, client client.Client, githubClient github.Client) secretUpdater {
+	return secretUpdater{
+		log:          log.WithName("secretUpdater"),
 		client:       client,
-		interval:     interval,
 		githubClient: githubClient,
+		processes:    map[string]*updateProcess{},
 	}
 }
 
-type SecretUpdater struct {
+type secretUpdater struct {
+	log          logr.Logger
 	client       client.Client
-	interval     time.Duration
 	githubClient github.Client
+	processes    map[string]*updateProcess
+}
+
+func (u secretUpdater) start(ctx context.Context, rp *meowsv1alpha1.RunnerPool) error {
+	rpNamespacedNameStr := types.NamespacedName{Namespace: rp.Namespace, Name: rp.Name}.String()
+	if _, ok := u.processes[rpNamespacedNameStr]; !ok {
+		process := newUpdateProcess(
+			u.log,
+			u.client,
+			u.githubClient,
+			rp.Spec.RepositoryName,
+			rpNamespacedNameStr,
+			types.NamespacedName{Namespace: rp.Namespace, Name: rp.GetRunnerSecretName()},
+		)
+		err := process.start(ctx)
+		if err != nil {
+			return err
+		}
+		u.processes[rpNamespacedNameStr] = process
+	}
+	return nil
+}
+
+func (u *secretUpdater) stop(ctx context.Context, rp *meowsv1alpha1.RunnerPool) error {
+	rpNamespacedNameStr := types.NamespacedName{Namespace: rp.Namespace, Name: rp.Name}.String()
+	if process, ok := u.processes[rpNamespacedNameStr]; ok {
+		if err := process.stop(ctx); err != nil {
+			return err
+		}
+		delete(u.processes, rpNamespacedNameStr)
+	}
+	return nil
+}
+
+type updateProcess struct {
+	log                  logr.Logger
+	client               client.Client
+	githubClient         github.Client
+	repositoryName       string
+	secretNamespacedName types.NamespacedName
+	env                  *well.Environment
+	cancel               context.CancelFunc
+
+	metrics       metricsSet
+	deleteMetrics func()
 }
 
-// Start implements Runnable.Start
-func (u SecretUpdater) Start(ctx context.Context) error {
-	ticker := time.NewTicker(u.interval)
-	defer ticker.Stop()
-	logger := log.FromContext(ctx).WithName("SecretUpdater")
+func newUpdateProcess(log logr.Logger, client client.Client, githubClient github.Client, repositoryName, rpNamespacedNameStr string, secretNamespacedName types.NamespacedName) *updateProcess {
+	return &updateProcess{
+		log:                  log.WithValues("RunnerPoolName", rpNamespacedNameStr),
```

Comment:
Please use the same field name.
https://github.com/cybozu-go/meows/blob/78b88a57f863794e032455e579fe30a78060db72/controllers/runnerpool_controller.go#L61

```suggestion
		log:                  log.WithValues("runnerpool", rpNamespacedNameStr),
```